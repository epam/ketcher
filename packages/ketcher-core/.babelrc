{
  "presets": ["@babel/env", "@babel/typescript"],
  "plugins": ["@babel/plugin-transform-runtime"]
}
