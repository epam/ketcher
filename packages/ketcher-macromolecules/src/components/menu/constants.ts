export const SELECT_SUBMENU_ID = 'select-submenu';
