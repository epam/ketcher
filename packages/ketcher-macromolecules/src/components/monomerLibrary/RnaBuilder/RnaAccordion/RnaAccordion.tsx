/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import { MonomerGroup } from 'components/monomerLibrary/monomerLibraryGroup';
import { useAppSelector } from 'hooks';
import { IconName } from 'ketcher-react';
import { useEffect, useState } from 'react';
import {
  MonomerCodeToGroup,
  MonomerGroupCodes,
  MonomerGroups,
} from 'src/constants';
import {
  getMonomerUniqueKey,
  selectFilteredMonomers,
  selectMonomerGroups,
  selectMonomersInCategory,
  selectUnsplitNucleotides,
  selectAmbiguousMonomersInCategory,
} from 'state/library';
import {
  DetailsContainer,
  DisabledArea,
  RnaAccordionContainer,
  StyledAccordion,
  StyledAccordionWrapper,
  StyledButton,
} from './styles';
import { Summary } from './Summary';
import {
  createNewPreset,
  RnaBuilderItem,
  RnaBuilderPresetsItem,
  selectActivePreset,
  selectActiveRnaBuilderItem,
  selectFilteredPresets,
  selectIsActivePresetNewAndEmpty,
  selectIsEditMode,
  setActivePresetMonomerGroup,
  setActiveRnaBuilderItem,
  setBaseValidations,
  setIsEditMode,
  setPhosphateValidations,
  setSugarValidations,
} from 'state/rna-builder';
import { useDispatch } from 'react-redux';
import { IRnaPreset } from '../types';
import {
  isAmbiguousMonomerLibraryItem,
  KetMonomerClass,
  MonomerItemType,
} from 'ketcher-core';
import {
  selectEditor,
  selectIsSequenceEditInRNABuilderMode,
} from 'state/common';
import { RnaPresetGroup } from 'components/monomerLibrary/RnaPresetGroup/RnaPresetGroup';
import { getValidations } from 'helpers/rnaValidations';

interface IGroupsDataItem {
  groupName: MonomerGroups | RnaBuilderPresetsItem;
  iconName: string;
  groups: {
    groupItems: IRnaPreset[] | MonomerItemType[];
    groupTitle?: string;
  }[];
}

export const RnaAccordion = ({ libraryName, duplicatePreset, editPreset }) => {
  const monomers = useAppSelector(selectFilteredMonomers);
  const items = selectMonomersInCategory(monomers, libraryName);
  const activeRnaBuilderItem = useAppSelector(selectActiveRnaBuilderItem);
  const activePreset = useAppSelector(selectActivePreset);
  const groups = selectMonomerGroups(items);
  const presets = useAppSelector(selectFilteredPresets);
  const nucleotideItems = selectUnsplitNucleotides(monomers);
  const nucleotideGroups = selectMonomerGroups(nucleotideItems);
  const isEditMode = useAppSelector(selectIsEditMode);
  const editor = useAppSelector(selectEditor);
  const isActivePresetNewAndEmpty = useAppSelector(
    selectIsActivePresetNewAndEmpty,
  );
  const isSequenceEditInRNABuilderMode = useAppSelector(
    selectIsSequenceEditInRNABuilderMode,
  );

  const [expandedAccordion, setExpandedAccordion] =
    useState<RnaBuilderItem | null>(activeRnaBuilderItem);
  const [newPreset, setNewPreset] = useState(activePreset);
  const [activeMonomerKey, setActiveMonomerKey] = useState('');

  const dispatch = useDispatch();
  const handleAccordionSummaryClick = (rnaBuilderItem: RnaBuilderItem) => {
    if (expandedAccordion === rnaBuilderItem) {
      setExpandedAccordion(null);
    } else {
      setExpandedAccordion(rnaBuilderItem);
      const { sugarValidations, phosphateValidations, baseValidations } =
        getValidations(newPreset, isEditMode);

      dispatch(setSugarValidations(sugarValidations));
      dispatch(setPhosphateValidations(phosphateValidations));
      dispatch(setBaseValidations(baseValidations));
    }
  };

  const groupsData: IGroupsDataItem[] = [
    {
      groupName: RnaBuilderPresetsItem.Presets,
      iconName: 'preset',
      groups: [{ groupItems: presets }],
    },
    {
      groupName: MonomerGroups.SUGARS,
      iconName: 'sugar',
      groups: groups.filter(
        (group) =>
          MonomerCodeToGroup[group.groupTitle as MonomerGroupCodes] ===
          MonomerGroups.SUGARS,
      ),
    },
    {
      groupName: MonomerGroups.BASES,
      iconName: 'base',
      groups: groups
        .filter(
          (group) =>
            MonomerCodeToGroup[group.groupTitle as MonomerGroupCodes] ===
            MonomerGroups.BASES,
        )
        .map((group) => {
          return {
            ...group,
            groupItems: group.groupItems.filter(
              (item) => item.props?.MonomerClass !== KetMonomerClass.RNA,
            ),
          };
        }),
    },
    {
      groupName: MonomerGroups.PHOSPHATES,
      iconName: 'phosphate',
      groups: groups.filter(
        (group) =>
          MonomerCodeToGroup[group.groupTitle as MonomerGroupCodes] ===
          MonomerGroups.PHOSPHATES,
      ),
    },
    {
      groupName: MonomerGroups.NUCLEOTIDES,
      iconName: 'nucleotide',
      groups: nucleotideGroups,
    },
  ];

  const selectItem = (monomer, groupName) => {
    setActiveMonomerKey(getMonomerUniqueKey(monomer));

    if (!isSequenceEditInRNABuilderMode && !isEditMode) {
      editor.events.selectMonomer.dispatch(monomer);
    }

    if (!isEditMode) {
      return;
    }

    const monomerClass = isAmbiguousMonomerLibraryItem(monomer)
      ? monomer.monomers[0].monomerItem.props.MonomerClass?.toLowerCase()
      : monomer.props.MonomerClass.toLowerCase();
    const currentPreset = {
      ...newPreset,
      [monomerClass]: monomer,
    };
    setNewPreset(currentPreset);
    dispatch(setActivePresetMonomerGroup({ groupName, groupItem: monomer }));
    dispatch(setActiveRnaBuilderItem(groupName));
  };

  useEffect(() => {
    setExpandedAccordion(activeRnaBuilderItem);
  }, [activeRnaBuilderItem]);

  useEffect(() => {
    dispatch(
      setActiveRnaBuilderItem(
        isEditMode && activePreset
          ? activeRnaBuilderItem
          : RnaBuilderPresetsItem.Presets,
      ),
    );
  }, [isEditMode]);

  const onClickNewPreset = () => {
    dispatch(createNewPreset());
    dispatch(setActiveRnaBuilderItem(RnaBuilderPresetsItem.Presets));
    dispatch(setIsEditMode(true));
  };

  return (
    <RnaAccordionContainer data-testid="rna-accordion">
      {groupsData.map((groupData) => {
        const expanded = expandedAccordion === groupData.groupName;
        const variantMonomers = selectAmbiguousMonomersInCategory(
          monomers,
          groupData.groupName as MonomerGroups,
        );
        const quantity = [...groupData.groups, ...variantMonomers].reduce(
          (acc, group) => acc + group.groupItems.length || 0,
          0,
        );
        const summary = (
          <Summary
            iconName={groupData.iconName as IconName}
            groupName={groupData.groupName}
            quantity={quantity}
            expanded={expanded}
          />
        );
        const details =
          groupData.groupName === RnaBuilderPresetsItem.Presets ? (
            <DetailsContainer>
              <StyledButton onClick={() => onClickNewPreset()}>
                New Preset
              </StyledButton>
              <RnaPresetGroup
                duplicatePreset={duplicatePreset}
                editPreset={editPreset}
                presets={presets}
              />
              {isEditMode && !isActivePresetNewAndEmpty && <DisabledArea />}
            </DetailsContainer>
          ) : (
            <DetailsContainer>
              <>
                {groupData.groups.map(({ groupItems, groupTitle }) => {
                  return (
                    <MonomerGroup
                      key={groupTitle}
                      title={
                        groupData.groups.length > 1 ? groupTitle : undefined
                      }
                      groupName={groupData.groupName as MonomerGroups}
                      items={groupItems as MonomerItemType[]}
                      selectedMonomerUniqueKey={activeMonomerKey}
                      onItemClick={(monomer) =>
                        selectItem(monomer, groupData.groupName)
                      }
                    />
                  );
                })}
                {variantMonomers.map((group) => {
                  return (
                    <MonomerGroup
                      key={group.groupTitle}
                      title={group.groupTitle}
                      items={group.groupItems}
                      libraryName={libraryName}
                      selectedMonomerUniqueKey={activeMonomerKey}
                      onItemClick={(monomer) =>
                        selectItem(monomer, groupData.groupName)
                      }
                    />
                  );
                })}
              </>
            </DetailsContainer>
          );
        const presetsExpanded =
          groupData.groupName === RnaBuilderPresetsItem.Presets && expanded;

        return presetsExpanded ? (
          <StyledAccordionWrapper key={groupData.groupName}>
            <StyledAccordion
              data-testid="styled-accordion"
              summary={summary}
              details={details}
              expanded={expanded}
              onSummaryClick={() =>
                handleAccordionSummaryClick(groupData.groupName)
              }
            />
          </StyledAccordionWrapper>
        ) : (
          <StyledAccordion
            key={groupData.groupName}
            data-testid="styled-accordion"
            dataTestIdDetails={`rna-accordion-details-${groupData.groupName}`}
            summary={summary}
            details={details}
            expanded={expanded}
            onSummaryClick={() =>
              handleAccordionSummaryClick(groupData.groupName)
            }
          />
        );
      })}
    </RnaAccordionContainer>
  );
};
