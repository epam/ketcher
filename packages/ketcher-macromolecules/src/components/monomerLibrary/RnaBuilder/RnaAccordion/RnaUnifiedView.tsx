import { useEffect, useState } from 'react';
import clsx from 'clsx';
import { Icon, IconName } from 'ketcher-react';
import { useAppSelector } from 'hooks';
import { MonomerGroup } from 'components/monomerLibrary/monomerLibraryGroup';
import { RnaPresetGroup } from 'components/monomerLibrary/RnaPresetGroup/RnaPresetGroup';
import {
  createNewPreset,
  recalculateRnaBuilderValidations,
  setActivePresetMonomerGroup,
  setActiveRnaBuilderItem,
  setIsEditMode,
  selectActivePreset,
  selectActiveRnaBuilderItem,
  selectFilteredPresets,
  selectIsActivePresetNewAndEmpty,
  selectIsEditMode,
  RnaBuilderPresetsItem,
  RnaBuilderItem,
} from 'state/rna-builder';
import {
  selectFilteredMonomers,
  selectMonomerGroups,
  selectMonomersInCategory,
  selectUnsplitNucleotides,
  selectAmbiguousMonomersInCategory,
  getMonomerUniqueKey,
} from 'state/library';
import {
  selectEditor,
  selectIsSequenceEditInRNABuilderMode,
} from 'state/common';
import {
  LibraryNameType,
  MonomerCodeToGroup,
  MonomerGroups,
} from 'src/constants';
import {
  IRnaPreset,
  isAmbiguousMonomerLibraryItem,
  KetMonomerClass,
} from 'ketcher-core';

import {
  RnaTabsContainer,
  RnaTab,
  RnaTabWrapper,
  CompactDetailsContainer,
  RnaAccordionContainer,
  DetailsContainer,
  DisabledArea,
  StyledAccordion,
  StyledAccordionWrapper,
  StyledButton,
} from './styles';
import { Summary } from './Summary';
import { useDispatch } from 'react-redux';

interface RnaUnifiedViewProps {
  view: 'tabs' | 'accordion';
  libraryName: LibraryNameType;
  duplicatePreset: (preset?: IRnaPreset) => void;
  editPreset: (preset: IRnaPreset) => void;
}

export const RnaUnifiedView = ({
  view,
  libraryName,
  duplicatePreset,
  editPreset,
}: RnaUnifiedViewProps) => {
  const dispatch = useDispatch();
  const monomers = useAppSelector(selectFilteredMonomers);
  const items = selectMonomersInCategory(monomers, libraryName);
  const activeRnaBuilderItem = useAppSelector(selectActiveRnaBuilderItem);
  const activePreset = useAppSelector(selectActivePreset);
  const groups = selectMonomerGroups(items);
  const presets = useAppSelector(selectFilteredPresets);
  const nucleotideItems = selectUnsplitNucleotides(monomers);
  const nucleotideGroups = selectMonomerGroups(nucleotideItems);
  const isEditMode = useAppSelector(selectIsEditMode);
  const editor = useAppSelector(selectEditor);
  const isActivePresetNewAndEmpty = useAppSelector(
    selectIsActivePresetNewAndEmpty,
  );
  const isSequenceEditInRNABuilderMode = useAppSelector(
    selectIsSequenceEditInRNABuilderMode,
  );

  const [expandedAccordion, setExpandedAccordion] =
    useState<RnaBuilderItem | null>(activeRnaBuilderItem);
  const [newPreset, setNewPreset] = useState(activePreset);
  const [activeMonomerKey, setActiveMonomerKey] = useState('');

  const handleAccordionSummaryClick = (rnaBuilderItem: RnaBuilderItem) => {
    if (expandedAccordion === rnaBuilderItem) {
      setExpandedAccordion(null);
    } else {
      setExpandedAccordion(rnaBuilderItem);
      dispatch(
        recalculateRnaBuilderValidations({ rnaPreset: newPreset, isEditMode }),
      );
    }
  };

  useEffect(() => {
    setExpandedAccordion(activeRnaBuilderItem);
  }, [activeRnaBuilderItem]);

  useEffect(() => {
    dispatch(
      setActiveRnaBuilderItem(
        isEditMode && activePreset
          ? activeRnaBuilderItem
          : RnaBuilderPresetsItem.Presets,
      ),
    );
  }, [isEditMode]);

  const onClickNewPreset = () => {
    dispatch(createNewPreset());
    dispatch(setActiveRnaBuilderItem(RnaBuilderPresetsItem.Presets));
    dispatch(setIsEditMode(true));
  };

  const groupsData = [
    {
      groupName: RnaBuilderPresetsItem.Presets,
      iconName: 'preset',
      groups: [{ groupItems: presets }],
    },
    {
      groupName: MonomerGroups.SUGARS,
      iconName: 'sugar',
      groups: groups.filter(
        (group) =>
          MonomerCodeToGroup[group.groupTitle as string] ===
          MonomerGroups.SUGARS,
      ),
    },
    {
      groupName: MonomerGroups.BASES,
      iconName: 'base',
      groups: groups
        .filter(
          (group) =>
            MonomerCodeToGroup[group.groupTitle as string] ===
            MonomerGroups.BASES,
        )
        .map((group) => ({
          ...group,
          groupItems: group.groupItems.filter(
            (item) => item.props?.MonomerClass !== KetMonomerClass.RNA,
          ),
        })),
    },
    {
      groupName: MonomerGroups.PHOSPHATES,
      iconName: 'phosphate',
      groups: groups.filter(
        (group) =>
          MonomerCodeToGroup[group.groupTitle as string] ===
          MonomerGroups.PHOSPHATES,
      ),
    },
    {
      groupName: MonomerGroups.NUCLEOTIDES,
      iconName: 'nucleotide',
      groups: nucleotideGroups,
    },
  ];

  const selectItem = (monomer: any, groupName: any) => {
    setActiveMonomerKey(getMonomerUniqueKey(monomer));

    if (!isSequenceEditInRNABuilderMode && !isEditMode) {
      editor.events.selectMonomer.dispatch(monomer);
    }

    if (!isEditMode) {
      return;
    }

    const monomerClass = isAmbiguousMonomerLibraryItem(monomer)
      ? monomer.monomers[0].monomerItem.props.MonomerClass?.toLowerCase()
      : monomer.props.MonomerClass.toLowerCase();
    const currentPreset = {
      ...newPreset,
      [monomerClass]: monomer,
    };
    setNewPreset(currentPreset);
    dispatch(setActivePresetMonomerGroup({ groupName, groupItem: monomer }));
    dispatch(setActiveRnaBuilderItem(groupName));
  };

  const renderTabs = () => (
    <>
      <RnaTabsContainer>
        {groupsData.map((groupData) => {
          const showCaption = groupData.groupName === activeRnaBuilderItem;
          const variantMonomers = selectAmbiguousMonomersInCategory(
            monomers,
            groupData.groupName as MonomerGroups,
          );
          const quantity = [...groupData.groups, ...variantMonomers].reduce(
            (acc, group) => acc + (group.groupItems.length || 0),
            0,
          );
          const caption = showCaption
            ? `${groupData.groupName} (${quantity})`
            : null;
          return (
            <RnaTabWrapper
              key={groupData.groupName}
              className={clsx(showCaption && 'rna-tab--selected')}
            >
              <RnaTab
                label={caption}
                title={groupData.groupName}
                selected={showCaption}
                icon={
                  <Icon
                    name={groupData.iconName as IconName}
                    onClick={() =>
                      dispatch(setActiveRnaBuilderItem(groupData.groupName))
                    }
                  />
                }
              />
            </RnaTabWrapper>
          );
        })}
      </RnaTabsContainer>
      {groupsData.map((groupData) => {
        if (groupData.groupName !== activeRnaBuilderItem) return null;
        const variantMonomers = selectAmbiguousMonomersInCategory(
          monomers,
          groupData.groupName as MonomerGroups,
        );
        const details =
          groupData.groupName === RnaBuilderPresetsItem.Presets ? (
            <DetailsContainer>
              <StyledButton onClick={onClickNewPreset}>New Preset</StyledButton>
              <RnaPresetGroup
                duplicatePreset={duplicatePreset}
                editPreset={editPreset}
                presets={presets}
              />
              {isEditMode && !isActivePresetNewAndEmpty && <DisabledArea />}
            </DetailsContainer>
          ) : (
            <DetailsContainer>
              <>
                {groupData.groups.map(({ groupItems, groupTitle }) => (
                  <MonomerGroup
                    key={groupTitle}
                    title={groupData.groups.length > 1 ? groupTitle : undefined}
                    groupName={groupData.groupName}
                    items={groupItems}
                    selectedMonomerUniqueKey={activeMonomerKey}
                    onItemClick={(monomer) =>
                      selectItem(monomer, groupData.groupName)
                    }
                  />
                ))}
                {variantMonomers.map((group) => (
                  <MonomerGroup
                    key={group.groupTitle}
                    title={group.groupTitle}
                    items={group.groupItems}
                    libraryName={libraryName}
                    selectedMonomerUniqueKey={activeMonomerKey}
                    onItemClick={(monomer) =>
                      selectItem(monomer, groupData.groupName)
                    }
                  />
                ))}
              </>
            </DetailsContainer>
          );

        const firstTabSelected =
          activeRnaBuilderItem === RnaBuilderPresetsItem.Presets;
        const lastTabSelected =
          activeRnaBuilderItem === MonomerGroups.NUCLEOTIDES;

        return (
          <CompactDetailsContainer
            className={clsx(
              firstTabSelected && 'first-tab',
              lastTabSelected && 'last-tab',
            )}
            key={groupData.groupName}
          >
            {details}
          </CompactDetailsContainer>
        );
      })}
    </>
  );

  const renderAccordion = () => (
    <>
      {groupsData.map((groupData) => {
        const expanded = expandedAccordion === groupData.groupName;
        const variantMonomers = selectAmbiguousMonomersInCategory(
          monomers,
          groupData.groupName as MonomerGroups,
        );
        const quantity = [...groupData.groups, ...variantMonomers].reduce(
          (acc, group) => acc + (group.groupItems.length || 0),
          0,
        );
        const summary = (
          <Summary
            iconName={groupData.iconName as IconName}
            groupName={groupData.groupName}
            quantity={quantity}
            expanded={expanded}
          />
        );
        const details =
          groupData.groupName === RnaBuilderPresetsItem.Presets ? (
            <DetailsContainer>
              <StyledButton onClick={onClickNewPreset}>New Preset</StyledButton>
              <RnaPresetGroup
                duplicatePreset={duplicatePreset}
                editPreset={editPreset}
                presets={presets}
              />
              {isEditMode && !isActivePresetNewAndEmpty && <DisabledArea />}
            </DetailsContainer>
          ) : (
            <DetailsContainer>
              <>
                {groupData.groups.map(({ groupItems, groupTitle }) => (
                  <MonomerGroup
                    key={groupTitle}
                    title={groupData.groups.length > 1 ? groupTitle : undefined}
                    groupName={groupData.groupName}
                    items={groupItems}
                    selectedMonomerUniqueKey={activeMonomerKey}
                    onItemClick={(monomer) =>
                      selectItem(monomer, groupData.groupName)
                    }
                  />
                ))}
                {variantMonomers.map((group) => (
                  <MonomerGroup
                    key={group.groupTitle}
                    title={group.groupTitle}
                    items={group.groupItems}
                    libraryName={libraryName}
                    selectedMonomerUniqueKey={activeMonomerKey}
                    onItemClick={(monomer) =>
                      selectItem(monomer, groupData.groupName)
                    }
                  />
                ))}
              </>
            </DetailsContainer>
          );
        return groupData.groupName === RnaBuilderPresetsItem.Presets &&
          expanded ? (
          <StyledAccordionWrapper key={groupData.groupName}>
            <StyledAccordion
              data-testid="styled-accordion"
              dataTestIdDetails={`rna-accordion-details-${groupData.groupName}`}
              summary={summary}
              details={details}
              expanded={expanded}
              onSummaryClick={() =>
                handleAccordionSummaryClick(groupData.groupName)
              }
            />
          </StyledAccordionWrapper>
        ) : (
          <StyledAccordion
            key={groupData.groupName}
            data-testid="styled-accordion"
            dataTestIdDetails={`rna-accordion-details-${groupData.groupName}`}
            summary={summary}
            details={details}
            expanded={expanded}
            onSummaryClick={() =>
              handleAccordionSummaryClick(groupData.groupName)
            }
          />
        );
      })}
    </>
  );

  return (
    <RnaAccordionContainer data-testid="rna-accordion">
      {view === 'tabs' ? renderTabs() : renderAccordion()}
    </RnaAccordionContainer>
  );
};
