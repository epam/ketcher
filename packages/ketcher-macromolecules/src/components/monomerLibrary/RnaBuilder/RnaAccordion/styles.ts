/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import styled from '@emotion/styled';
import { Accordion, Button } from 'ketcher-react';
import { Tab } from '@mui/material';

export const RnaAccordionContainer = styled.div({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'flex-start',
  overflow: 'hidden',
  height: '100%',
});

export const StyledAccordion = styled(Accordion)({
  minHeight: '32px',
});

export const StyledAccordionWrapper = styled.div({
  flexGrow: 2,
  minHeight: '32px',
});

export const DetailsContainer = styled.div({
  position: 'relative',
  width: '100%',
  display: 'flex',
  flexDirection: 'column',
  gap: '8px',
  justifyContent: 'start',
  padding: '8px',
});

export const StyledButton = styled(Button)((props) => {
  return {
    backgroundColor: props.theme.ketcher.color.button.transparent.active,
    color: props.theme.ketcher.color.text.light,
    borderColor: props.theme.ketcher.color.text.light,
  };
});

export const DisabledArea = styled.div({
  width: '100%',
  height: '100%',
  backgroundColor: '#eff2f594',
  position: 'absolute',
  top: 0,
  left: 0,
});

export const RnaTabsContainer = styled.div`
  display: flex;
  gap: 8px;
  padding: 6px 8px 0;
`;

export const RnaTab = styled(Tab)`
  height: 24px;
  min-height: 24px;
  min-width: 24px;
  display: flex;
  flex-direction: row;
  gap: 4px;
  align-items: center;
  justify-content: center;
  padding: 4px;
  font-weight: 400;
  font-size: 10px;
  border-radius: 4px;
  background-color: white;
  opacity: 1;
  text-transform: none;

  > svg {
    height: 16px;
    width: 16px;
    color: #b4b9d6;

    &.MuiTab-iconWrapper {
      margin: 0;
    }
  }
`;
