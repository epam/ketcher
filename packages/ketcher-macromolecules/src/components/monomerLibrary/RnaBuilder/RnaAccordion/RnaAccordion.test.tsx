import { render, screen } from '@testing-library/react';
import { RnaAccordion } from 'components/monomerLibrary/RnaBuilder/RnaAccordion/RnaAccordion';
import { getMonomerUniqueKey } from 'state/library';
import { MonomerItemType, Struct } from 'ketcher-core';
import { MONOMER_TYPES } from '../../../../constants';

const duplicatePreset = jest.fn();
const editPreset = jest.fn();
describe('Test Rna Accordion component', () => {
  it('should render', () => {
    render(
      withThemeAndStoreProvider(
        <RnaAccordion
          libraryName={MONOMER_TYPES.RNA}
          duplicatePreset={duplicatePreset}
          editPreset={editPreset}
        />,
      ),
    );

    const rnaAccordion = screen.getByTestId('rna-accordion');

    expect(rnaAccordion).toMatchSnapshot();
  });

  it('should show number of items in section', () => {
    const monomerData = {
      label: '',
      struct: new Struct(),
      props: {
        Name: "1',2'-Di-Deoxy-Ribose",
        MonomerType: 'RNA',
        MonomerName: '12ddR',
        MonomerCode: '',
        MonomerNaturalAnalogCode: 'P',
        BranchMonomer: 'false',
        MonomerCaps: { R1: 'H' },
      },
    };
    render(
      withThemeAndStoreProvider(
        <RnaAccordion
          libraryName={MONOMER_TYPES.RNA}
          duplicatePreset={duplicatePreset}
          editPreset={editPreset}
        />,
        {
          library: {
            searchFilter: '',
            favorites: {},
            monomers: [monomerData],
          },
        },
      ),
    );
    const peptidesSummary = screen.getByTestId('summary-Phosphates');

    expect(peptidesSummary).toHaveTextContent('Phosphates (1)');

    const monomerCard = screen.getByTestId(
      getMonomerUniqueKey(monomerData as MonomerItemType),
    );
    expect(monomerCard).toBeVisible();
  });
});
