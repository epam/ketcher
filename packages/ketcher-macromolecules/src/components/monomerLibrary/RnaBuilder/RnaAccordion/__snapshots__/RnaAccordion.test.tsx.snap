// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rna Accordion component should render 1`] = `
<div
  class="css-r9woj4"
  data-testid="rna-accordion"
>
  <div
    class="css-11h3u53"
  >
    <div
      class="css-1svedvx"
    >
      <div
        class="css-s4s9u"
      >
        <div
          class="css-75ynye"
          data-testid="summary-Presets"
        >
          <svg
            class="css-1qr96ji"
            fill="none"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="currentColor"
              height="5.333"
              rx="0.5"
              width="5.333"
              x="2"
              y="1.327"
            />
            <rect
              fill="currentColor"
              height="4.667"
              rx="0.5"
              transform="rotate(45 -7.332 9.589)"
              width="4.667"
            />
            <rect
              fill="currentColor"
              height="5.333"
              rx="2.667"
              width="5.333"
              x="9.336"
              y="1.327"
            />
          </svg>
          <span
            class="css-lqq9ua"
          >
            Presets
             (
            0
            )
          </span>
          <svg
            class="css-1qr96ji"
            fill="none"
            height="48"
            viewBox="0 0 48 48"
            width="48"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 33.23L9.782 19.012l2.828-2.828L24 27.574l11.388-11.39 2.83 2.828L24 33.23z"
              fill="#231F20"
            />
          </svg>
        </div>
      </div>
      <div
        class="css-2kwivx"
      >
        <div
          class="css-9bccu"
        >
          <button
            class="css-1f4nv0y"
          >
            New Preset
          </button>
          <div
            class="css-1jn33hb"
            data-testid="rna-preset-group"
          >
            <div
              class="css-14bnc38"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="css-1svedvx"
  >
    <div
      class="css-s4s9u"
    >
      <div
        class="css-75ynye"
        data-testid="summary-Sugars"
      >
        <svg
          class="css-1qr96ji"
          fill="none"
          height="20"
          stroke-width="2"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="currentColor"
            height="18"
            rx="1"
            width="18"
            x="1"
            y="1"
          />
        </svg>
        <span
          class="css-lqq9ua"
        >
          Sugars
           (
          0
          )
        </span>
        <svg
          class="css-x2ejol"
          fill="none"
          height="48"
          viewBox="0 0 48 48"
          width="48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M24 33.23L9.782 19.012l2.828-2.828L24 27.574l11.388-11.39 2.83 2.828L24 33.23z"
            fill="#231F20"
          />
        </svg>
      </div>
    </div>
    <div
      class="css-4wpvad"
      data-testid="rna-accordion-details-Sugars"
    >
      <div
        class="css-9bccu"
      />
    </div>
  </div>
  <div
    class="css-1svedvx"
  >
    <div
      class="css-s4s9u"
    >
      <div
        class="css-75ynye"
        data-testid="summary-Bases"
      >
        <svg
          class="css-1qr96ji"
          fill="none"
          height="25"
          stroke-width="2"
          transform="translate(-3)"
          viewBox="0 0 26 25"
          width="26"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="currentColor"
            height="15.013"
            rx="1"
            transform="rotate(45 13 1.414)"
            width="15.013"
            x="13"
            y="1.414"
          />
          <rect
            height="15.013"
            rx="1"
            stroke="currentColor"
            transform="rotate(45 13 1.414)"
            width="15.013"
            x="13"
            y="1.414"
          />
        </svg>
        <span
          class="css-lqq9ua"
        >
          Bases
           (
          0
          )
        </span>
        <svg
          class="css-x2ejol"
          fill="none"
          height="48"
          viewBox="0 0 48 48"
          width="48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M24 33.23L9.782 19.012l2.828-2.828L24 27.574l11.388-11.39 2.83 2.828L24 33.23z"
            fill="#231F20"
          />
        </svg>
      </div>
    </div>
    <div
      class="css-4wpvad"
      data-testid="rna-accordion-details-Bases"
    >
      <div
        class="css-9bccu"
      />
    </div>
  </div>
  <div
    class="css-1svedvx"
  >
    <div
      class="css-s4s9u"
    >
      <div
        class="css-75ynye"
        data-testid="summary-Phosphates"
      >
        <svg
          class="css-1qr96ji"
          fill="none"
          height="24"
          stroke-width="2"
          transform="translate(-2)"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="currentColor"
            height="18"
            rx="9"
            width="18"
            x="3"
            y="3"
          />
          <rect
            height="18"
            rx="9"
            stroke="currentColor"
            width="18"
            x="3"
            y="3"
          />
        </svg>
        <span
          class="css-lqq9ua"
        >
          Phosphates
           (
          0
          )
        </span>
        <svg
          class="css-x2ejol"
          fill="none"
          height="48"
          viewBox="0 0 48 48"
          width="48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M24 33.23L9.782 19.012l2.828-2.828L24 27.574l11.388-11.39 2.83 2.828L24 33.23z"
            fill="#231F20"
          />
        </svg>
      </div>
    </div>
    <div
      class="css-4wpvad"
      data-testid="rna-accordion-details-Phosphates"
    >
      <div
        class="css-9bccu"
      />
    </div>
  </div>
</div>
`;
